server/mqtt.go

```diff
@@ -4958,9 +4947,75 @@ func isMQTTReservedSubscription(subject string) bool {
 	return false
 }
 
+func sparkbReplaceDeathTimestamp(msg []byte) []byte {
+	const VARINT = 0
+	const TIMESTAMP = 1
+
+	orig := msg
+	buf := bytes.NewBuffer(make([]byte, 0, len(msg)+16)) // 16 bytes should be enough if we need to add a timestamp
+	writeDeathTimestamp := func() {
+		// [tck-id-conformance-mqtt-aware-ndeath-timestamp] A Sparkplug Aware
+		// MQTT Server MAY replace the timestamp of NDEATH messages. If it does,
+		// it MUST set the timestamp to the UTC time at which it attempts to
+		// deliver the NDEATH to subscribed clients
+		//
+		// sparkB spec: 6.4.1. Google Protocol Buffer Schema
+		//      optional uint64 timestamp = 1; // Timestamp at message sending time
+		ts := uint64(time.Now().UnixMilli())
+		buf.Write(protoEncodeVarint(TIMESTAMP<<3 | VARINT))
+		buf.Write(protoEncodeVarint(ts))
+	}
+
+	for len(msg) > 0 {
+		numField, typ, size, err := protoScanField(msg)
+		if err != nil {
+			return orig
+		}
+		if typ != VARINT || numField != TIMESTAMP {
+			// Add the field as is
+			buf.Write(msg[:size])
+			msg = msg[size:]
+			continue
+		}
+
+		writeDeathTimestamp()
+
+		// Add the rest of the message as is, we are done
+		buf.Write(msg[size:])
+		return buf.Bytes()
+	}
+
+	// Add timestamp if we did not find one.
+	writeDeathTimestamp()
+
+	return buf.Bytes()
+}
+
 // Common function to mqtt delivery callbacks to serialize and send the message
 // to the `cc` client.
-func (c *client) mqttEnqueuePublishMsgTo(cc *client, sub *subscription, pi uint16, qos byte, dup, retain bool, topic, msg []byte) {
+func (c *client) mqttEnqueuePublishMsgTo(cc *client, sub *subscription, pi uint16, qos byte, dup bool, topic, msg []byte) {
+	// [tck-id-conformance-mqtt-aware-nbirth-mqtt-retain] A Sparkplug Aware
+	// MQTT Server MUST make NBIRTH messages available on the topic:
+	// $sparkplug/certificates/namespace/group_id/NBIRTH/edge_node_id with
+	// the MQTT retain flag set to true
+	//
+	// [tck-id-conformance-mqtt-aware-dbirth-mqtt-retain] A Sparkplug Aware
+	// MQTT Server MUST make DBIRTH messages available on the topic:
+	// $sparkplug/certificates/namespace/group_id/DBIRTH/edge_node_id/device_id
+	// with the MQTT retain flag set to true
+	//
+	// $sparkplug/certificates messages are sent as NATS messages, so we
+	// need to add the retain flag when sending them to MQTT ciients.
+
+	retain := false
+	isBirth, isDeath, isCertificate := sparkbParseBirthDeathTopic(topic)
```

Comment:
I hope that the check (we start with a bytes.HasPrefix()) is fast enough to return false for non sparkb birt/death messages to reduce perf impact.

Reply:
Right. Like you said in the main PR, we can put this on an `if config option`. Otherwise I don't see how we can avoid prefix-matching all incoming MQTT messages for DEATH (even if we could somehow short circuit the BIRTH handling).

I will compare and paste the before/after benchmark results.

Reply:
But also, in the other PR, you did this equivalent check only in QoS0 callback, where putting the check here applies to all QoS's. Again, not that it matters much performance-wise for QoS > 0 anyway (since they are comparatively slow).

Reply:
Right. I moved it here because we have to parse for DEATH anyway. Will add the QOS check, too.

Reply:
• [tck-id-message-flow-edge-node-birth-publish-will-message-qos] The Edge Node’s MQTT Will
Message’s MQTT QoS MUST be 1.

I changed the test but checking the DEATH qos proved more difficult since we may deliver it as a QoS0 message.

Now checking BIRTH to be qos0